rust/benchmarks/src/main.rs

```diff
@@ -0,0 +1,139 @@
+use std::process;
+use std::time::Instant;
+
+use arrow::datatypes::{DataType, Field, Schema};
+use arrow::util::pretty;
+use datafusion::error::Result;
+use datafusion::execution::context::ExecutionContext;
+
+use structopt::StructOpt;
+
+#[derive(Debug, StructOpt)]
+#[structopt(name = "Benchmarks", about = "Apache Arrow Rust Benchmarks.")]
+struct Opt {
+    /// Activate debug mode to see query results
+    #[structopt(short, long)]
+    debug: bool,
+
+    /// Number of iterations of each test run
+    #[structopt(short = "i", long = "iterations", default_value = "3")]
+    iterations: usize,
+
+    /// Batch size when reading CSV or Parquet files
+    #[structopt(short = "s", long = "batch-size", default_value = "4096")]
+    batch_size: usize,
+
+    /// Path to data files
+    #[structopt(parse(from_os_str), required = true, short = "p", long = "path")]
+    path: PathBuf,
+
+    /// File format: `csv` or `parquet`
+    #[structopt(short = "f", long = "format", default_value = "csv")]
+    file_format: String,
+}
+
+fn main() -> Result<()> {
+    let opt = Opt::from_args();
+    println!("Running benchmarks with the following options: {:?}", opt);
+
+    let mut ctx = ExecutionContext::new();
+
+    let path = opt.path.to_str().unwrap();
+
+    match opt.file_format.as_str() {
+        "csv" => ctx.register_csv("tripdata", path, &nyctaxi_schema(), true),
+        "parquet" => ctx.register_parquet("tripdata", path)?,
+        other => {
+            println!("Invalid file format '{}'", other);
+            process::exit(-1);
+        }
+    }
+
+    datafusion_sql_benchmarks(&mut ctx, opt.iterations, opt.batch_size, opt.debug)
+}
+
+fn datafusion_sql_benchmarks(
+    ctx: &mut ExecutionContext,
+    iterations: usize,
+    batch_size: usize,
+    debug: bool,
+) -> Result<()> {
+    let mut queries = HashMap::new();
+    queries.insert("fare_amt_by_passenger", "SELECT passenger_count, MIN(fare_amount), MIN(fare_amount), SUM(fare_amount) FROM tripdata GROUP BY passenger_count");
+    for (name, sql) in &queries {
+        println!("Executing '{}'", name);
+        for i in 0..iterations {
+            let start = Instant::now();
+            execute_sql(ctx, sql, batch_size, debug)?;
+            println!(
+                "Query '{}' iteration {} took {} ms",
+                name,
+                i,
+                start.elapsed().as_millis()
+            );
+        }
+    }
+    Ok(())
+}
+
+fn execute_sql(
+    ctx: &mut ExecutionContext,
+    sql: &str,
+    batch_size: usize,
+    debug: bool,
+) -> Result<()> {
+    let plan = ctx.create_logical_plan(sql)?;
+    let plan = ctx.optimize(&plan)?;
+    if debug {
+        println!("Optimized logical plan:\n{:?}", plan);
+    }
+    let physical_plan = ctx.create_physical_plan(&plan, batch_size)?;
+    let result = ctx.collect(physical_plan.as_ref())?;
+    if debug {
+        pretty::print_batches(&result)?;
+    }
+    Ok(())
+}
+
+fn nyctaxi_schema() -> Schema {
```

Comment:
Do you have a normalized dataset?

Reply:
No, I've just been testing queries against a single year so far and this is the correct schema for that year.

If/when we extend these benchmarks to query across multiple years we would need to improve this. 

DataFusion doesn't yet support schema merging.